smart-home-provider/cloud/auth-provider.js

```diff
@@ -304,6 +302,53 @@ Auth.registerAuth = function(app) {
 // &grant_type=authorization_code
 
 
+/**
+ * Decode authorization into client_id and client_secret.
+ * @return {{}}
+ * {
+ *   client_id: "CLIENT_ID",
+ *   client_secret: "CLIENT_SECRET"
+ * }
+ */
+function getHeaderAuthorization(req){
+  let headerAuthor = {};
+    try{
+      if(!!req.headers.authorization){
+        let authorizationString = Buffer.from(req.headers.authorization.split(" ")[1], 'base64').toString();
+        if(!!authorizationString){
+          let authorizationArray = authorizationString.split(':');
+          headerAuthor = {
+            client_id: authorizationArray[0],
+            client_secret: authorizationArray[1]
+          }
+        }
+      }
+    }catch(e){
+      return {};
+    }
+    return headerAuthor;
+}
+
+/**
+ * Get client's key from the following order:
+ *   - Request's query string
+ *   - Request's body
+ *   - Request's header (from `authorization`)
+ * @return string
+ */
+function getClientValue(key, req){
+  if(!!req.query[key])
```

Comment:
Why do we need to double negate the value?

Reply:
This is just to make sure that we're parsing to boolean as [this answer](https://stackoverflow.com/a/784946/5561570) describes.

Reply:
If `req.query[key]` is `undefined`, it is already a falsy value